tests/components/onedrive/test_backup.py

```diff
@@ -0,0 +1,326 @@
+) -> None:
+    """Test agent upload backup."""
+    client = await hass_client()
+    test_backup = AgentBackup.from_dict(BACKUP_METADATA)
+
+    with (
+        patch(
+            "homeassistant.components.backup.manager.BackupManager.async_get_backup",
+        ) as fetch_backup,
+        patch(
+            "homeassistant.components.backup.manager.read_backup",
+            return_value=test_backup,
+        ),
+        patch("pathlib.Path.open") as mocked_open,
+        patch("homeassistant.components.onedrive.backup.UPLOAD_CHUNK_SIZE", 3),
+    ):
+        mocked_open.return_value.read = Mock(side_effect=[b"test", b""])
+        fetch_backup.return_value = test_backup
+        resp = await client.post(
+            f"/api/backup/upload?agent_id={DOMAIN}.{mock_config_entry.title}",
+            data={"file": StringIO("test")},
+        )
+
+    assert resp.status == 201
+    assert f"Uploading backup {test_backup.backup_id}" in caplog.text
+    mock_drive_items.create_upload_session.post.assert_called_once()
+    mock_drive_items.patch.assert_called_once()
+    assert mock_adapter.send_async.call_count == 2
+
+
+async def test_broken_upload_session(
+    hass_client: ClientSessionGenerator,
+    caplog: pytest.LogCaptureFixture,
+    mock_drive_items: MagicMock,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test broken upload session."""
+    client = await hass_client()
+    test_backup = AgentBackup.from_dict(BACKUP_METADATA)
+
+    mock_drive_items.create_upload_session.post = AsyncMock(return_value=None)
+
+    with (
+        patch(
+            "homeassistant.components.backup.manager.BackupManager.async_get_backup",
+        ) as fetch_backup,
+        patch(
+            "homeassistant.components.backup.manager.read_backup",
+            return_value=test_backup,
+        ),
+        patch("pathlib.Path.open") as mocked_open,
+    ):
+        mocked_open.return_value.read = Mock(side_effect=[b"test", b""])
+        fetch_backup.return_value = test_backup
+        resp = await client.post(
+            f"/api/backup/upload?agent_id={DOMAIN}.{mock_config_entry.title}",
+            data={"file": StringIO("test")},
+        )
+
+    assert resp.status == 201
+    assert "Failed to start backup upload" in caplog.text
+
+
+async def test_agents_download(
+    hass_client: ClientSessionGenerator,
+    mock_drive_items: MagicMock,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test agent download backup."""
+    mock_drive_items.get = AsyncMock(
+        return_value=DriveItem(description=escape(dumps(BACKUP_METADATA)))
+    )
+    client = await hass_client()
+    backup_id = BACKUP_METADATA["backup_id"]
+
+    resp = await client.get(
+        f"/api/backup/download/{backup_id}?agent_id={DOMAIN}.{mock_config_entry.title}"
+    )
+    assert resp.status == 200
+    assert await resp.content.read() == b"backup data"
+    mock_drive_items.content.get.assert_called_once()
+
+
+async def test_error_wrapper(
```

Comment:
I would suggest renaming this to `test_delete_error`, since error_wrapper is an internal thing.
Alternatively, we could parameterize the test to check for all the operations (i.e. define the `delete`, `get`, etc methods as params)